### PR TITLE
add basic search

### PR DESCRIPTION
**Pull request by sufikaur**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,7 +20,8 @@
     "limiter": "^1.1.5",
     "mariadb": "^2.4.1",
     "memory-cache": "^0.2.0",
-    "node-fetch": "^2.6.0"
+    "node-fetch": "^2.6.0",
+    "swagger-ui-dist": "^3.18.1"
```
Comment (tsingh777) [inline package.json:24]: should be separate PR 😄

File: api/routes/index.js [hunk 2]
```diff
@@ -27,18 +27,31 @@ route.get('/hello', limiter.rate250, (req, res) => res.send('Hello World!'));
 // healthcheck
 route.get('/health', limiter.rate250, healthcheck);
 
-// all albums
-route.get('/albums', limiter.rate250, allAlbums);
+// ALBUMS
+/**
+ * /albums : all albums
+ * /albums?nameincl={nameIncludes} : Search for album that matches this name (fuzzy match)
```
Comment (tsingh777) [inline api/routes/index.js:33]: call this `name`

File: api/controllers/album.js [hunk 3]
```diff
@@ -38,3 +75,58 @@ export async function byAlbumID(req, res) {
...
+  let q = '';
+  const basicQuery = `SELECT ${cols} FROM Album WHERE 1=1`;
+  // parent
+  if (parentID) {
+    q += ` AND Album.Parent = ${parentID}`;
+    params.push(parentID);
+  }
+  // name
+  if (name) {
+    q += ` AND Album.Title LIKE '%${name}%'`;
```
Comment (tsingh777) [inline api/controllers/album.js:103]: change this to 
```
if (name) {
    q += ' AND Album.Title LIKE ?';
    params.push(`%${name}%`);
  }
```

File: api/controllers/album.js [hunk 4]
```diff
@@ -38,3 +75,58 @@ export async function byAlbumID(req, res) {
...
+    params.push(name);
+  }
+  // tag
+  if (tag) {
+    q += ` AND Album.Tags LIKE '%${tag}%'`;
```
Comment (tsingh777) [inline api/controllers/album.js:108]: move % to parameters and out of query.

File: api/controllers/album.js [hunk 5]
```diff
@@ -38,3 +75,58 @@ export async function byAlbumID(req, res) {
...
+    params.push(tag);
+  }
+  // date
+  if (date) {
+    q += ` AND Album.Updated > '${date}'`;
+    params.push(date);
+  }
+  // keyword
+  if (keyword) {
+    q += ` AND Album.Keywords LIKE '%${keyword}%'`;
```
Comment (tsingh777) [inline api/controllers/album.js:118]: move % to parameters and out of query.

File: api/controllers/album.js [hunk 6]
```diff
@@ -1,13 +1,44 @@
...
+export async function albumsBy(req, res) {
   let conn;
+  const nameIncludes = req.query.nameincl;
+  // lint/prettier keeps trying to {tag} but everything breaks then
+  //const { tag } = req.query;
+
+  const tag = req.query.tag;
+  const parentID = parseInt(req.query.parentid, 10);
+  const date = req.query.date;
+  const { keyword } = req.query;
```
Comment (tsingh777) [inline api/controllers/album.js:23]: combine the other props from req.query into 1 destructured statement.

File: api/controllers/album.js [hunk 7]
```diff
@@ -1,13 +1,44 @@
...
-export async function allAlbums(req, res) {
+
+import { query } from 'express';
+
+/**
+ * /albums : all albums
+ * /albums?nameincl={name} : Search for album that matches this name (fuzzy match)
+ * /albums?tag={tag} : Search for albums that have these tags (fuzzy matches)
+ * /albums?keyword={keyword} : Search for albums that have these keywords (fuzzy matches)
+ * /albums?parentid={parentid} : Search for albums that has this albumId as a parent.
```
Comment (tsingh777) [inline api/controllers/album.js:11]: make this camelCase `parantId`

File: api/controllers/album.js [hunk 8]
```diff
@@ -1,13 +1,44 @@
...
+ * /albums?updatedafter={date} : Search for albums that were last updated after this date.
```
Comment (tsingh777) [inline api/controllers/album.js:12]: simplify to `updated`

File: api/controllers/album.js [hunk 9]
```diff
@@ -38,3 +75,58 @@ export async function byAlbumID(req, res) {
...
+  if (albumID) {
+    q += ` AND Album.ID = ${albumID}`;
+    params.push(albumID);
+  }
+  // // should we do this?
+  // else {
+  //   q += ` AND Parent IS NULL`;
+  // }
+  // return an object with a query and its parameters
+  return [`${basicQuery + q};`, params];
```
Comment (tsingh777) [inline api/controllers/album.js:131]: better to return object instead of array. 
```
return { query:`${basicQuery + q};`, params };
```
then you can access 
`const { query, params } = getAlbumQuery(...)`.

That way order doesn't matter.

File: api/controllers/album.js [hunk 10]
```diff
@@ -38,3 +75,58 @@ export async function byAlbumID(req, res) {
...
+ * @param {String} name - string for fuzzy search
+ * @param {String} tag - tag for fuzzy search
+ * @param {Int} parentID - parentID for exact search
+ * @param {String} date - updated after given date
+ * @param {String} keyword - keyword for fuzzy search
+ * @param {Int} albumID - albumID for exact search
+ *
+ * @return {Object[String, Array]} = template statement, parameters for template statement
+ */
+function getAlbumQuery(name, tag, parentID, date, keyword, albumID) {
```
Comment (tsingh777) [inline api/controllers/album.js:91]: change the input to an object so order can be ignored. 
```
function getAlbumQuery({ name, tag, parentID, date, keyword, albumID }) {
```

you would call it like 
`getAlbumQuery({name: 'foo', tag: 'bar' }); // input object can any of the props.`

File: api/controllers/album.js [hunk 11]
```diff
@@ -1,13 +1,44 @@
 // function: allAlbums
 // sends all parent albums
-export async function allAlbums(req, res) {
+
```
Comment (tsingh777) [inline api/controllers/album.js:3]: similar changes needed in the tracks file.

File: api/controllers/album.js [hunk 12]
```diff
@@ -1,13 +1,44 @@
...
+ * /albums?parentid={parentid} : Search for albums that has this albumId as a parent.
+ * /albums?updatedafter={date} : Search for albums that were last updated after this date.
+ */
+export async function albumsBy(req, res) {
   let conn;
+  const nameIncludes = req.query.nameincl;
+  // lint/prettier keeps trying to {tag} but everything breaks then
+  //const { tag } = req.query;
+
+  const tag = req.query.tag;
```
Comment (tsingh777) [inline api/controllers/album.js:20]: the params that are strings could be sanitized. somethings like 
- check if its a string,  
- if the string length is greater than 0. 
- trim whitespace from the ends of the string
- Convert case ? 

(@ManjotS can a `WHERE` clause that uses `LIKE` be case in-sensitive)?

File: api/controllers/album.js (hunk 6 above)
Comment (sufikaur) [reply api/controllers/album.js:23]: what does this mean?

File: api/controllers/track.js [hunk 13]
```diff
@@ -49,3 +90,92 @@ export async function byTrackID(req, res) {
...
+    params.push(`%${name}%`);
+  }
+  // length
+  if (length) {
+    q += ` AND Length > ?`;
+    params.push(`${length}`);
+  }
+  // type
+  if (type) {
+    q += ` AND Type = ?`;
```
Comment (tsingh777) [inline api/controllers/track.js:149]: single quote

File: api/controllers/track.js [hunk 14]
```diff
@@ -49,3 +90,92 @@ export async function byTrackID(req, res) {
...
+    params.push(`${type}`);
+  }
+  // track id
+  if (trackID) {
+    q += ` AND ID = ?`;
```
Comment (tsingh777) [inline api/controllers/track.js:154]: single quote

File: api/controllers/track.js [hunk 15]
```diff
@@ -49,3 +90,92 @@ export async function byTrackID(req, res) {
...
+
+/**
+ * function: getAlbumQuery
+ * creates query and parameters for an album query to the database
+ * currently does not work with prepared statements
+ *
+ * @param {String} albumID - string for fuzzy search
+ * @param {String} artistID - tag for fuzzy search
+ * @param {Int} name - parentID for exact search
+ * @param {String} length - updated after given date
```
Comment (tsingh777) [inline api/controllers/track.js:102]: should you break this up into `longerThan` and `shorterThan`? Can come back to this since that will require some validation making sure longerThan is less than shorterThan.

File: api/controllers/album.js [hunk 16]
```diff
@@ -38,3 +73,59 @@ export async function byAlbumID(req, res) {
...
+ * @param {String} tag - tag for fuzzy search
+ * @param {Int} parentID - parentID for exact search
+ * @param {String} date - updated after given date
+ * @param {String} keyword - keyword for fuzzy search
+ * @param {Int} albumID - albumID for exact search
+ *
+ * @return {Object[String, Array]} = template statement, parameters for template statement
+ */
+function getAlbumQuery({ name, tag, parentID, updated, keyword, albumID }) {
+  const cols = `ID, Title, Parent, Tags, Keywords, Artists, Updated`;
```
[the remark this answers is not in the document]
Comment (sufikaur) [reply api/controllers/album.js:90]: it has issues merging into the one that does which is why I left it there.

File: api/controllers/track.js (hunk 15 above)
Comment (sufikaur) [reply api/controllers/track.js:102]: ok- lets come back to this

File: api/controllers/album.js [hunk 17]
```diff
@@ -1,13 +1,32 @@
...
+ * /albums?tag={tag} : Search for albums that have these tags (fuzzy matches)
+ * /albums?keyword={keyword} : Search for albums that have these keywords (fuzzy matches)
+ * /albums?parentID={parentID} : Search for albums that has this albumId as a parent.
+ * /albums?updated={updated} : Search for albums that were last updated after this date.
+ */
+export async function albumsBy(req, res) {
   let conn;
+  const { name } = req.query;
+  const { tag } = req.query;
+  const { keyword } = req.query;
```
Comment (tsingh777) [inline api/controllers/album.js:16]: can combine these.

File: api/controllers/track.js [hunk 18]
```diff
@@ -1,24 +1,30 @@
...
+export async function tracksBy(req, res) {
   let conn;
-  let q;
-  const albumid = parseInt(req.query.albumid, 10);
-  let result;
+  const albumID = parseInt(req.query.albumID, 10);
+  const artistID = parseInt(req.query.artistid, 10);
+  const { name } = req.query;
+  const { length } = req.query;
+  const { type } = req.query;
```
Comment (tsingh777) [inline api/controllers/track.js:15]: can combine.

File: api/controllers/album.js (hunk 17 above)
Comment (sufikaur) [reply api/controllers/album.js:16]: I keep doing it! Prettier/lint keeps separating them.

File: api/controllers/album.js [hunk 19]
```diff
@@ -38,3 +57,53 @@ export async function byAlbumID(req, res) {
...
+  }
+  // date
+  if (updated) {
+    q += ' AND Album.Updated > ?';
+    params.push(`${updated}`);
+  }
+  // keyword
+  if (keyword) {
+    q += ' AND Album.Keywords LIKE ?';
+    params.push(`%${keyword}`);
```
Comment (tsingh777) [inline api/controllers/album.js:101]: missing `%` at the end.

File: api/controllers/album.js [hunk 20]
```diff
@@ -1,13 +1,32 @@
 // function: allAlbums
 // sends all parent albums
-export async function allAlbums(req, res) {
+
+/**
+ * /albums : all albums
+ * /albums?name={name} : Search for album that matches this name (fuzzy match)
```
Comment (tsingh777) [inline api/controllers/album.js:6]: Tempted to say lets change this to title.

File: api/controllers/album.js [hunk 21]
```diff
@@ -1,13 +1,32 @@
...
+ * /albums?tag={tag} : Search for albums that have these tags (fuzzy matches)
```
Comment (tsingh777) [inline api/controllers/album.js:7]: thinking this should be `tags` and should take an array (comma separated list)

File: api/controllers/album.js [hunk 22]
```diff
@@ -1,13 +1,32 @@
...
+ * /albums?keyword={keyword} : Search for albums that have these keywords (fuzzy matches)
```
Comment (tsingh777) [inline api/controllers/album.js:8]: same for this.

File: api/controllers/album.js [hunk 23]
```diff
@@ -1,13 +1,32 @@
...
+ * /albums?parentID={parentID} : Search for albums that has this albumId as a parent.
```
Comment (tsingh777) [inline api/controllers/album.js:9]: maybe this should be just `parent`.
